src/org/exist/collections/Collection.java

```diff
@@ -1096,6 +1096,20 @@ public void removeCollection(final DBBroker broker, final XmldbURI name) throws
         }
     }
 
+    public void removeResource(Txn txn, DBBroker broker, DocumentImpl doc) throws PermissionDeniedException, LockException, IOException, TriggerException {
+        if (doc == null) return;
+
+        if (doc.getCollection() != this) {
```

Comment:
because `removeResource` method is public and `DocumentImpl` come in from outside. Also, because there is a time interval between getting `DocumentImpl` and passing it to that method when it have to lock. The getting `DocumentImpl` with read lock lead to deadlock because it break locking logic that collection must be locked 1st and document 2nd.